feat: Godot debug symbols guide (#12843)
getsentry/sentry-docs: docs/platforms/godot/configuration/stack-traces.mdx

---

> title: Readable Stack Traces

Author: Feel free to propose a better name.

</Alert>

## Getting Godot Source

Start in a terminal with Git available. Clone Godot Engine repository and switch to your preferred version tag (or branch):
```bash
git clone https://github.com/godotengine/godot
cd godot
git checkout 4.3-stable
```

## Compiling Templates

To compile the Godot release export template with debug symbols, run the following command:

```bash
scons platform=windows target=template_release production=yes debug_symbols=yes separate_debug_symbols=yes
```

<Expandable title="Options explained">

- `platform`: target platform (e.g., "windows", "linux", or "macos")
- `target`: compilation target (e.g., "editor" or "template_release")
- `production`: sets build defaults for production use
- `debug_symbols`: includes debugging symbols in the build
- `separate_debug_symbols`: extracts symbols into a separate file

For more information, run: `scons --help`

</Expandable>

If all goes well, the export template files should appear as the result of the build process in the `bin/` directory inside the Godot source tree. When using MSVC on Windows, you will find `.exe` files alongside `.pdb` files, which contain debug information.

<Alert>

> Learn more about debug formats in [Debug Information Files](/platforms/native/data-management/debug-files/).

Author: Linking to Native page, as it's more likely to stay up-to-date on this topic.

Author: It's still unclear to me what is the consensus on this.

This will configure your project to use the custom export template with the debug symbols you compiled earlier.

![Export Dialog](./imgs/godot_export_dialog.png)

You should now be able to export your project using this template. Just make sure to uncheck `Export with Debug` in the **Export Project...** dialog, as you will need to compile and use the `template_debug` for that option.

<Alert>

If you need to `Export with Debug`, you can compile a debug template using the `target=template_debug` option. Debug templates enable debugging features in your export, such as breakpoints and debugger support.

</Alert>

## Uploading Debug Symbols

In order to get readable stack traces in Sentry, you also need to upload your debug symbols. The easiest way to do it is by using Sentry CLI tools. These tools allow you to upload the necessary debug information, making the stack traces useful for debugging.

<Expandable title="Installing Sentry CLI">

On Windows, if you have `scoop` installed:
```PowerShell
scoop install sentry-cli
```

On macOS, if you have Homebrew installed:
```bash
brew install getsentry/tools/sentry-cli
```

> On Arch Linux (btw):

Reviewer: ```suggestion
On Arch Linux:
```

```bash
pacman -S sentry-cli
```

For other methods, refer to [Sentry CLI Installation](/cli/installation/).

> </Expandable>

Reviewer: We should link to this: https://docs.sentry.io/cli/installation/

If we're going to add some commands here, we should at least render as a tab component.
e.g: https://github.com/getsentry/sentry-docs/blob/c9290bd0ba43c2ff081cf6ad0541016cec7c0939/docs/platforms/dotnet/index.mdx?plain=1#L42-L48

I think the problem here is the text, between the ``` blocks.

Author: We do link to that page at the bottom. 
Do you mean, we should link to **installation** page and avoid adding any command instructions?

Author: I've refactored the instructions as tab component in 66780ce0153d279ab898f680725b63252e09adc2

Log in to Sentry via the CLI by running the following command:

```bash
sentry-cli login
```

It will prompt you to create an auth token in your web browser. Follow the instructions, generate the token, and then paste it into the command-line prompt when asked.

To upload debug symbols to Sentry using `sentry-cli`, run the following command:

```bash
sentry-cli debug-files upload --org <YOUR_ORGANIZATION> --project <YOUR_PROJECT> ./bin
```

Replace `<YOUR_ORGANIZATION>` and `<YOUR_PROJECT>` with your actual organization and project slugs or IDs. This will upload the files with debug information from the `./bin` directory to Sentry.

Example output:

```PowerShell

> $ sentry-cli debug-files upload --org sentry-sdks --project sentry-godot ./bin

Reviewer: ```suggestion
$ sentry-cli debug-files upload --org my-sentry-org --project my-game ./bin
```

---
title: Readable Stack Traces
description: "Learn how to get readable stack traces in Sentry with SDK for Godot Engine."
---

> Debug information files allow Sentry to extract stack traces and provide additional information about crash reports. In order to get readable stack traces in Sentry when your game crashes, you need to export your project using an export template that contains debug information. For the official builds, Godot Engine only provides templates without debug symbols. This guide covers how to create such templates and use them to export your project with full debug information available in Sentry.

Reviewer: ```suggestion
Debug information files allow Sentry to extract stack traces and provide additional information from crash reports. In order to get stack traces that contain function names, line numbers and file paths, Sentry needs to have these debug information files available when processing crash dumps. For that, when you need to export your Godot project using an export template that contains debug information. The official Godot Engine builds only provide templates without debug symbols. This guide covers how to create such templates and use them to export your project with full debug information available in Sentry.
```

> Debug information files allow Sentry to extract stack traces and provide additional information from crash reports. In order to get stack traces that contain function names, line numbers and file paths, Sentry needs to have these debug information files available when processing crash dumps. For that, you need to export your Godot project using an export template that contains debug information. The official Godot Engine builds only provide templates without debug symbols. This guide covers how to create such templates and use them to export your project with full debug information available in Sentry.

Reviewer: ```suggestion
[The official Godot Engine builds only provide templates without debug information files](https://github.com/godotengine/godot-proposals/issues/1342). This guide covers how to create such templates and use them to export your project with full debug information available in Sentry.

Debug information files allow Sentry to extract stack traces and provide additional information from crash reports. In order to get stack traces that contain function names, line numbers and file paths, Sentry needs to have these debug information files available when processing the crash dumps. For that, you need to export your Godot project using an export template that contains debug information. 
```

> Debug information files allow Sentry to extract stack traces and provide additional information from crash reports. In order to get stack traces that contain function names, line numbers and file paths, Sentry needs to have these debug information files available when processing crash dumps. For that, when you need to export your Godot project using an export template that contains debug information. The official Godot Engine builds only provide templates without debug symbols. This guide covers how to create such templates and use them to export your project with full debug information available in Sentry.

Author: ```suggestion
Debug information files allow Sentry to extract stack traces and provide additional information from crash reports. In order to get stack traces that contain function names, line numbers and file paths, Sentry needs to have these debug information files available when processing crash dumps. For that, you need to export your Godot project using an export template that contains debug information. The official Godot Engine builds only provide templates without debug symbols. This guide covers how to create such templates and use them to export your project with full debug information available in Sentry.
```

Ensure you have a compatible C++ compiler. On Windows, for instance, you can use [Visual Studio Community 2022](https://visualstudio.microsoft.com/) with the C++ workload installed, GCC on Linux and Clang on macOS.

Here are a few ways to install the **SCons** build tool:

```PowerShell {tabTitle:Windows}
# If you have `scoop` installed:
scoop install scons
```

```bash {tabTitle:macOS}
# If you have Homebrew installed:
brew install scons
```

```bash {tabTitle:Arch Linux}
pacman -S scons
```

```bash {tabTitle:Ubuntu/Debian}
sudo apt install scons
```

```bash {tabTitle:Using Python}
# Install with existing Python installation:
python -m pip install scons

# Upgrade:
python -m pip install --upgrade scons
```

<Alert>

For more information, refer to [Building from Source](https://docs.godotengine.org/en/stable/contributing/development/compiling/index.html) in the official Godot documentation. It provides detailed instructions for compiling Godot on different platforms and with different tools. This guide offers streamlined instructions tailored for our use case, and may omit some details you require.

</Alert>

## Getting Godot Source

> Start in a terminal with Git available. Clone Godot Engine repository and switch to your preferred version tag (or branch):

Reviewer: ```suggestion
Start in a terminal with Git available. Clone Godot Engine repository and switch to your [preferred version tag](https://github.com/godotengine/godot/releases) (or branch):
```

```bash
git clone https://github.com/godotengine/godot
cd godot

> git checkout 4.3-stable

Reviewer: ```suggestion
git checkout 4.4-stable
```

For more information, run: `scons --help`

</Expandable>

If all goes well, the export template files should appear as the result of the build process in the `bin/` directory inside the Godot source tree. When using MSVC on Windows, you will find `.exe` files alongside `.pdb` files, which contain debug information.

<Alert>

Learn more about debug formats in [Debug Information Files](/platforms/native/data-management/debug-files/).

</Alert>

## Exporting Project

Now you can export your project using your custom template. In the Godot **Export** dialog window, follow these steps:

1. Select or add an export preset.
2. Enable the **Advanced Options** toggle.
3. In the **Options** tab, under **Custom Template -> Release**, assign your custom template executable file.

This will configure your project to use the custom export template with the debug symbols you compiled earlier.

![Export Dialog](./imgs/godot_export_dialog.png)

You should now be able to export your project using this template. Just make sure to uncheck `Export with Debug` in the **Export Project...** dialog, as you will need to compile and use the `template_debug` for that option.

<Alert>

> If you need to `Export with Debug`, you can compile a debug template using the `target=template_debug` option. Debug templates enable debugging features in your export, such as breakpoints and debugger support.

Reviewer: ```suggestion
If you need to `Export with Debug`, you can compile a debug template using the `target=template_debug` option. Debug templates enable debugging features in your export, such as breakpoints and debugger support but perform slower than a release build, therefore these are not advised when shipping the game for end users.
```

</Alert>

## Uploading Debug Symbols

> In order to get readable stack traces in Sentry, you also need to upload your debug symbols. The easiest way to do it is by using Sentry CLI tools. These tools allow you to upload the necessary debug information, making the stack traces useful for debugging.

Reviewer: ```suggestion
In order to get readable stack traces in Sentry, you also need to upload your debug symbols. The easiest way to do it is by using [Sentry CLI](https://docs.sentry.io/cli/). This tool allows you to upload the necessary debug information, making the stack traces useful for debugging.
```

<Expandable title="Installing Sentry CLI">

```PowerShell {tabTitle:Windows}
# On Windows, if you have `scoop` installed:
scoop install sentry-cli
```

```bash {tabTitle:macOS}
# On macOS, if you have Homebrew installed:
brew install getsentry/tools/sentry-cli
```

```bash {tabTitle:Arch Linux}
pacman -S sentry-cli
```

```npm {tabTitle:npm}
npm install @sentry/cli
```

For other methods, refer to [Sentry CLI Installation](/cli/installation/).

</Expandable>

Log in to Sentry via the CLI by running the following command:

```bash
sentry-cli login
```

It will prompt you to create an auth token in your web browser. Follow the instructions, generate the token, and then paste it into the command-line prompt when asked.

To upload debug symbols to Sentry using `sentry-cli`, run the following command:

```bash

> sentry-cli debug-files upload --org <YOUR_ORGANIZATION> --project <YOUR_PROJECT> ./bin

Reviewer 2: ```suggestion
sentry-cli debug-files upload --org ___ORG_SLUG___ --project ___PROJECT_SLUG___ ./bin
```
This will allow users to select their org and project in the docs and get a ready-to-paste snippet

Author: Applied

```

> Replace `<YOUR_ORGANIZATION>` and `<YOUR_PROJECT>` with your actual organization and project slugs or IDs. This will upload the files with debug information from the `./bin` directory to Sentry.

Reviewer 2: ```suggestion
This will upload the files with debug information from the `./bin` directory to Sentry.
```

Author: Applied
